onnxruntime/core/session/onnxruntime_c_api.cc

```diff
@@ -498,32 +481,66 @@ ORT_API_STATUS_IMPL(OrtGetStringTensorContent, _In_ const OrtValue* value,
   API_IMPL_END
 }
 
-ORT_API_STATUS_IMPL(OrtTensorProtoToOrtValue, _Inout_ OrtAllocator* allocator,
-                    const void* input, int input_len, _Out_ OrtValue** out) {
+#define ORT_C_API_RETURN_IF_ERROR(expr)                 \
+  do {                                                  \
+    auto _status = (expr);                              \
+    if ((!_status.IsOK())) return ToOrtStatus(_status); \
+  } while (0)
+
+ORT_API_STATUS_IMPL(OrtTensorProtoToOrtValue, _In_ const void* input, int input_len,
+                    _In_opt_ const ORTCHAR_T* input_file_path, _Inout_ void* preallocated, size_t preallocated_size,
+                    _Out_ OrtValue** out, _Out_ OrtDeleter** deleter) {
   API_IMPL_BEGIN
-  std::shared_ptr<onnxruntime::IAllocator> allocator_ = std::make_shared<onnxruntime::AllocatorWrapper>(allocator);
+  OrtAllocatorInfo* cpuAllocatorInfo;
+  auto st = OrtCreateAllocatorInfo("Cpu", OrtDeviceAllocator, 0, OrtMemTypeDefault, &cpuAllocatorInfo);
+  if (st != nullptr) return st;
   ::ONNX_NAMESPACE::TensorProto proto;
   if (!proto.ParseFromArray(input, input_len)) {
     return OrtCreateStatus(ORT_FAIL, "parse input tensor proto failed");
   }
   std::unique_ptr<MLValue> value = std::make_unique<MLValue>();
-  Status st = onnxruntime::utils::TensorProtoToMLValue(proto, allocator_, nullptr, 0, *value);
-  if (!st.IsOK())
-    return ToOrtStatus(st);
+  std::unique_ptr<OrtDeleter> del = std::make_unique<OrtDeleter>();
+  auto status =
+      utils::TensorProtoToMLValue(Env::Default(), input_file_path, proto,
+                                  MemBuffer(preallocated, preallocated_size, *cpuAllocatorInfo), *value, *del);
+  OrtReleaseAllocatorInfo(cpuAllocatorInfo);
+  if (!status.IsOK()) {
+    return ToOrtStatus(status);
+  }
   *out = reinterpret_cast<OrtValue*>(value.release());
+  if (del->f != nullptr) {
+    *deleter = del.release();
+  } else
+    *deleter = nullptr;
   return nullptr;
   API_IMPL_END
 }
 
+ORT_API_STATUS_IMPL(OrtGetTensorMemSizeInBytesFromTensorProto, _In_ const void* input, int input_len, size_t alignment,
+                    size_t* out) {
+  API_IMPL_BEGIN
+  ::ONNX_NAMESPACE::TensorProto proto;
+  if (!proto.ParseFromArray(input, input_len)) {
+    return OrtCreateStatus(ORT_FAIL, "parse input tensor proto failed");
+  }
+  switch (alignment) {
+    case 0:
+      ORT_C_API_RETURN_IF_ERROR(utils::GetSizeInBytesFromTensorProto<0>(proto, out));
+      break;
+    case 256:
+      ORT_C_API_RETURN_IF_ERROR(utils::GetSizeInBytesFromTensorProto<256>(proto, out));
+      break;
+    default:
+      return OrtCreateStatus(ORT_INVALID_ARGUMENT, "invalid alignment");
```

Comment:
>invalid alignment [](start = 52, length = 17)

Ideally the error message helps the user directly fix the issue, so better to say what values are valid as well.